### PR TITLE
Only lookup once, not once per thread.

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst or @toregge or @havardpe PR

### CODE REVIEW

**Comment by geirst** [comment]
Unit test in searchlib fails:
`381/739 Test #386: searchlib_prod_features_test_app .................................***Failed    0.22 sec
prod_features_test: info:  running test suite 'prod_features_test'
prod_features_test: info:  flushed 70 passed check(s) from thread 'master' (prod_features.cpp:87)
prod_features_test: info:  flushed 34 passed check(s) from thread 'master' (prod_features.cpp:88)
prod_features_test: info:  flushed 22 passed check(s) from thread 'master' (prod_features.cpp:89)
/source/searchlib/src/tests/features/prod_features_test.sh: line 4:  5265 Segmentation fault      (core dumped) VESPA_LOG_TARGET=file:vlog2.txt $VALGRIND ./searchlib_prod_features_test_app`

**Review by geirst — COMMENTED** [review]
👍 But check the failing unit test first

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp [hunk 1]
```diff
@@ -413,10 +413,20 @@ createTensorAttributeExecutor(const IAttributeVector *attribute, const vespalib:
...
 
-search::fef::FeatureExecutor &
-AttributeBlueprint::createExecutor(const search::fef::IQueryEnvironment &env, vespalib::Stash &stash) const
+void
+AttributeBlueprint::prepareSharedState(const fef::IQueryEnvironment & env, fef::IObjectStore &) const
+{
+    //Ensure that we remember it the whole query
+    const IAttributeVector * attribute = env.getAttributeContext().getAttribute(_attrName);
+    if (_attribute == nullptr) {
+        _attribute = attribute;
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/attributefeature.cpp]: The query environment attribute context will make sure that this pointer is valid during query execution. This pointer will be used for multiple queries, which seems very unsafe and undefined. Please put this pointer in the object store, which stores per-query state.

**Comment by havardpe** [comment]
This seems to introduce an issue similar to the dotproduct issue not so long ago.

**Comment by baldersheim** [comment]
Then the problem was that the attribute name was overrideable query time. Here it is not.
However I see that there is an issue if prepareSharedState is not always called.

**Comment by havardpe** [comment]
The problem is following a pointer that may no longer lead you to an attribute vector. You basically have a raw pointer to something that is kept alive by shared pointers. The attribute context in the query environment will keep a guard during query execution, but when the query completes all bets are off.

**Comment by baldersheim** [comment]
But the attribute itself will have at least the same life span as the Blueprint due to reconfig is the only time things should change. But I agree that is not guaranteed in this code.

**Comment by toregge** [comment]
For imported attribute vectors, the pointer will become invalid when the attribute context is destroyed.
For non-imported attribute vectors, the read guard is dropped when the attribute context is destroyed.

**Comment by baldersheim** [comment]
Ah, reverting.

**Pull request merged**